fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/BulkDataUtils.java

```diff
@@ -155,41 +156,89 @@ private static int getFhirResourceFromBufferReader(BufferedReader resReader, int
-        } catch (Exception ioe) {
-            logger.warning("readFhirResourceFromObjectStore: " + "Error proccesing file " + itemName + " - " + ioe.getMessage());
-            exported = 0;
+           int numOfLinesToSkip, List<Resource> fhirResources, boolean isReuseInput, ImportTransientUserData transientUserData) {
+        int exported = 0;
+        if (isReuseInput) {
+            if (transientUserData.getBufferReader() == null) {
+                S3Object item = cosClient.getObject(new GetObjectRequest(bucketName, itemName));
+                S3ObjectInputStream s3InStream = item.getObjectContent();
+                transientUserData.setInputStream(s3InStream);
+                BufferedReader resReader = new BufferedReader(new InputStreamReader(s3InStream));
+                transientUserData.setBufferReader(resReader);
+            }
+            try {
+                exported = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), 0, fhirResources);
+            } catch (Exception ioe) {
+                logger.warning("readFhirResourceFromObjectStore: " + "Error proccesing file " + itemName + " - " + ioe.getMessage());
+                exported = 0;
+            }
+
+        } else {
+            S3Object item = cosClient.getObject(new GetObjectRequest(bucketName, itemName));
+            try (S3ObjectInputStream s3InStream = item.getObjectContent();
+                 BufferedReader resReader = new BufferedReader(new InputStreamReader(s3InStream))) {
+                exported = getFhirResourceFromBufferReader(resReader, numOfLinesToSkip, fhirResources);
+                // Notify s3 client to abort and prevent the server from keeping on sending data.
+                s3InStream.abort();
+            } catch (Exception ioe) {
+                logger.warning("readFhirResourceFromObjectStore: " + "Error proccesing file " + itemName + " - " + ioe.getMessage());
+                exported = 0;
+            }
         }
+
         return exported;
     }
 
 
-    public static int readFhirResourceFromLocalFile(String filePath, int numOfLinesToSkip, List<Resource> fhirResources) {
-        int exported;
-        try (BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath))) {
-            exported = getFhirResourceFromBufferReader(resReader, numOfLinesToSkip, fhirResources);
-        } catch (Exception ioe) {
-            logger.warning("readFhirResourceFromLocalFile: " + "Error proccesing file " + filePath + " - " + ioe.getMessage());
-            exported = 0;
+    public static int readFhirResourceFromLocalFile(String filePath, int numOfLinesToSkip, List<Resource> fhirResources,
+            boolean isReuseInput, ImportTransientUserData transientUserData) {
+        int exported = 0;
+        if (isReuseInput) {
+            try {
+                if (transientUserData.getBufferReader() == null) {
+                    BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath));
```

Comment:
I think there should be some defensive programming here to validate the path is acceptable.  This may be best at the REST layer.

Reply:
yes, make sense for me. in the job side, if got a non-exising file, the job will still succeeded but with 0 imported.  throwing exception can cause the whole job fails even if only one non-existing file there.